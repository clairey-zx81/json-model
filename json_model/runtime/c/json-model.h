#ifndef JSON_MODEL_H_INCLUDED
#define JSON_MODEL_H_INCLUDED

#define _GNU_SOURCE  // for qsort_r?
#include <stdbool.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#include <jansson.h>

/*
 * build generated API names
 */
#ifndef CHECK_FUNCTION_NAME
#warning defining CHECK_FUNCTION_NAME macro
#define CHECK_FUNCTION_NAME check_model
#endif  // !CHECK_FUNCTION_NAME

// branch hints
#define likely(x)       __builtin_expect(!!(x), 1)
#define unlikely(x)     __builtin_expect(!!(x), 0)

// inlining hints
#if defined(NO_INLINE_HINT)
#define INLINE
#else
#if defined(__clang__)
#define INLINE __attribute__((always_inline))
#elif defined(__GNUC__)
#define INLINE inline
// #elif defined(_MSC_VER)
#else
#define INLINE
#endif  // clang/gcc/...
#endif  // NO_INLINE_HINT

#define concat(prefix, suffix) prefix ## suffix
#define newname(prefix, suffix) concat(prefix, suffix)

#define CHECK_map_size newname(CHECK_FUNCTION_NAME, _map_size)
#define CHECK_map_tab newname(CHECK_FUNCTION_NAME, _map_tab)
#define CHECK_fun newname(CHECK_FUNCTION_NAME, _map)
#define CHECK_init newname(CHECK_FUNCTION_NAME, _init)
#define CHECK CHECK_FUNCTION_NAME
#define CHECK_free newname(CHECK_FUNCTION_NAME, _free)

// set on init
extern char * jm_version_string;

#define array_length(arr) (sizeof(arr) / sizeof((arr)[0]))
typedef int (*jm_cmp_fun_t)(const void *, const void *);

/*
 * JSON API extensions
 */

// keep object prop/vals as an array for sorting and fast get.
typedef struct {
    const char *prop;
    const json_t *val;
} jm_propval_t;

extern int jm_propval_cmp(const jm_propval_t *, const jm_propval_t *);

extern bool jm_json_is_scalar(const json_t *);
extern int jm_json_cmp(const json_t *, const json_t *);
extern int jm_json_array_cmp(const json_t *, const json_t *);
extern int jm_json_object_cmp(const json_t *, const json_t *);
extern bool jm_json_array_unique(const json_t *);

/*
 * reporting
 */
// linked list of report entries
typedef struct jm_report_entry {
  const char *message;
  char *path;
  struct jm_report_entry *prev;
} jm_report_entry_t;

typedef struct {
  jm_report_entry_t *entry;
} jm_report_t;

// path segment, name == NULL means use index
typedef struct jm_path {
    const char *name;
    size_t index;
    struct jm_path *prev;  // link to previous (in stack) segment
    struct jm_path *next;  // used for backtacking
} jm_path_t;

extern void jm_report_add_entry(jm_report_t* rep, const char *msg, jm_path_t *path);
extern void jm_report_free_entries(jm_report_t *rep);

typedef bool (*jm_check_fun_t)(const json_t *, jm_path_t *, jm_report_t *);

extern bool jm_check_fun_string(jm_check_fun_t, const char *, jm_path_t *, jm_report_t *);
extern bool jm_array_is_unique(const json_t *, jm_path_t *, jm_report_t *);

/*
 * property mapping management
 */
typedef struct {
    const char *name;
    jm_check_fun_t val_check;
} jm_propmap_t;

extern void jm_sort_propmap(jm_propmap_t *, int);
extern jm_check_fun_t jm_search_propmap(const char *, const jm_propmap_t *, int);

/*
 * set of scalar constants
 */
typedef enum {
    cst_is_null,
    cst_is_bool,
    cst_is_integer,
    cst_is_float,
    cst_is_string
    // cst_is_array + const json_t * + free
    // cst_is_object + const json_t * + free
} jm_constant_tag;

typedef struct
{
    jm_constant_tag tag;
    union {
        bool b;
        int64_t i;
        double f;
        const char *s;
    } val;
} jm_constant_t;

extern bool jm_set_cst(jm_constant_t *, const json_t *);
extern void jm_dbg_cst(const jm_constant_t *, const char *);
extern void jm_sort_cst(jm_constant_t *, size_t);
extern bool jm_search_cst(const jm_constant_t *, const jm_constant_t *, size_t);

/*
 * constant mapping management for discriminant optimization
 */
typedef struct {
    jm_constant_t cst;
    jm_check_fun_t check_val;
} jm_constmap_t;

extern void jm_sort_constmap(jm_constmap_t *, size_t);
extern jm_check_fun_t jm_search_constmap(const jm_constant_t *, const jm_constmap_t *, size_t);

/*
 * Miscellaneous support functions and related data
 */
extern bool jm_is_valid_date_slow(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_date_fast(const char *, jm_path_t *, jm_report_t *);
extern bool (*jm_is_valid_date)(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_time(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_datetime(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_uuid(const char *, jm_path_t *, jm_report_t *);
extern size_t jm_str_len(const char *);
extern size_t jm_any_len(json_t *);
extern bool jm_is_valid_regex_slow(const char *, bool, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_regex_fast(const char *, bool, jm_path_t *, jm_report_t *);
extern bool (*jm_is_valid_regex)(const char *, bool, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_url(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_email(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_json(const char *, jm_path_t *, jm_report_t *);

typedef enum {
    op_eq,
    op_ne,
    op_le,
    op_lt,
    op_ge,
    op_gt
} jm_constraint_op_t;

extern bool
jm_check_constraint(const json_t *, jm_constraint_op_t, const jm_constant_t *,
                    jm_path_t *, jm_report_t *);

/*
 * Shared high-level entry point
 */
extern bool jm_generic_entry(const char *(*)(void), jm_check_fun_t (*)(const char *),
                             const json_t *, const char *, bool *, char **);

/*
 * Generated stuff
 */
// low-level interface
extern const size_t CHECK_map_size;
extern jm_propmap_t CHECK_map_tab[];
extern jm_check_fun_t CHECK_fun(const char *);
// high-level interface
extern const char *CHECK_init(void);
extern bool CHECK(const json_t *, const char *, bool *, char **);
extern void CHECK_free(void);

#endif  // JSON_MODEL_H_INCLUDED
