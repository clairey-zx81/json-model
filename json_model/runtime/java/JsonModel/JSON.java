package JsonModel;

import java.util.List;
import java.util.Map;
import java.util.Iterator;
import java.io.Reader;

/**
 * Convenient abstraction for JSON libraries such as GSON and Jackson.
 */
public abstract class JSON<T>
{
    public static class Exception extends java.lang.Exception
    {
        public Exception(String s) {
            super(s);
        }
    }

    public abstract T fromJSON(String json) throws Exception;
    public abstract T fromJSON(Reader reader) throws Exception;
    public abstract String toJSON(T o) throws Exception;
    public abstract boolean isNull(T o);
    public abstract boolean isBoolean(T o);
    public abstract boolean asBoolean(T o);
    public abstract boolean isInteger(T o);
    public abstract int asInteger(T o);
    public abstract long asLong(T o);
    public abstract boolean isDouble(T o);
    public abstract double asDouble(T o);

    public boolean isNumber(T o)
    {
        return isInteger(o) || isDouble(o);
    }

    public double asNumber(T o)
    {
        return isDouble(o) ? asDouble(o) : (double) asLong(o);
    }

    public abstract boolean isString(T o);
    public abstract String asString(T o);
    public abstract boolean isArray(T o);
    public abstract int arrayLength(T o);
    public abstract T arrayItem(T o, int index);
    public abstract Iterator<? extends T> arrayIterator(T o);
    public abstract boolean isObject(T o);
    public abstract int objectSize(T o);
    public abstract T objectValue(T o, String prop);
    public abstract Iterator<String> objectIterator(T o);
}
