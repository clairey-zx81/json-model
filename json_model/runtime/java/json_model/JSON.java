package json_model;

import java.util.List;
import java.util.Map;
import java.util.Iterator;
import java.io.Reader;

/**
 * Convenient abstraction for JSON libraries such as GSON and Jackson.
 *
 * This interface allows to run a compiled checker with different JSON libraries.
 */
public abstract class JSON<T>
{

    /** All JSON types */
    static public final int NULL = 1;
    static public final int BOOLEAN = 2;
    static public final int INTEGER = 3;
    static public final int NUMBER = 4;
    static public final int STRING = 5;
    static public final int ARRAY = 6;
    static public final int OBJECT = 7;

    /** Internal exception for JSON errors */
    public static class Exception extends java.lang.Exception
    {
        public Exception(String s) {
            super(s);
        }
    }

    /** Convert string to JSON thing */
    public abstract T fromJSON(String json) throws Exception;

    /** No exception version, for generated json */
    public T safeJSON(String json)
    {
        try {
            return fromJSON(json);
        }
        catch (Exception e) {
            throw new Error("unexpected exception while converting from json: " + e);
        }
    }

    /** Convert text flow to JSON thing */
    public abstract T fromJSON(Reader reader) throws Exception;

    /** Generate JSON string from Java string */
    public T strToJSON(String s)
    {
        return (T) s;
    }

    /** Convert JSON thing to a serialized JSON string */
    public abstract String toJSON(T o) throws Exception;

    /** Tell whether JSON thing is "null" */
    public abstract boolean isNull(T o);

    /** Tell whether JSON thing is a "boolean" */
    public abstract boolean isBoolean(T o);

    /** Extract boolean value from JSON thing */
    public abstract boolean asBoolean(T o);

    /** Tell whether JSON thing is an "integer" */
    public abstract boolean isInteger(T o);

    /** Extract int value from JSON thing */
    public abstract int asInteger(T o);

    /** Extract long value from JSON thing */
    public abstract long asLong(T o);

    /** Tell whether JSON thing is a "double" */
    public abstract boolean isDouble(T o);

    /** Extract double value from JSON thing */
    public abstract double asDouble(T o);

    /** Tell whether JSON thing is a "number" */
    public boolean isNumber(T o)
    {
        return isInteger(o) || isDouble(o);
    }

    /** Extract number value from JSON thing */
    public double asNumber(T o)
    {
        return isDouble(o) ? asDouble(o) : (double) asLong(o);
    }

    /** Tell whether JSON thing is a "string" */
    public abstract boolean isString(T o);

    /** Extract string value from JSON thing */
    public abstract String asString(T o);

    /** Tell whether JSON thing is a scalar */
    public boolean isScalar(T o)
    {
        return isString(o) || isNumber(o) || isBoolean(o) || isNull(o);
    }

    /** Tell whether JSON thing is an array */
    public abstract boolean isArray(T o);

    /** Get array length */
    public abstract int arrayLength(T o);

    /** Get array item at index */
    public abstract T arrayItem(T o, int index);

    /** Iterate over array items */
    public abstract Iterator<? extends T> arrayIterator(T o);

    /** Extract JSON array as a Java array */
    public Object[] asArray(T o)
    {
        int length = arrayLength(o);
        Object[] array = new Object[length];
        for (int i = 0; i < length; i++)
            array[i] = arrayItem(o, i);
        return array;
    }

    /** Tell whether JSON thing is an object */
    public abstract boolean isObject(T o);

    /** Get object size (aka number of properties) */
    public abstract int objectSize(T o);

    /** Get value for an object property */
    public abstract T objectValue(T o, String prop);

    /** Tell whether JSON thing has a property */
    public abstract boolean objectHasProp(T o, String prop);

    /** Iterate over a JSON object properties */
    public abstract Iterator<String> objectIterator(T o);

    /** Get JSON thing type */
    public int type(T o)
    {
        if (isNull(o))
            return NULL;
        if (isBoolean(o))
            return BOOLEAN;
        if (isInteger(o))
            return INTEGER;
        if (isDouble(o))
            return NUMBER;
        if (isString(o))
            return STRING;
        if (isArray(o))
            return ARRAY;
        if (isObject(o))
            return OBJECT;
        throw new Error("unexpected object: " + o);
    }
}
