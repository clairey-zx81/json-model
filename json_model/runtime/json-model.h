#ifndef JSON_MODEL_H_INCLUDED
#define JSON_MODEL_H_INCLUDED

#define _GNU_SOURCE  // for qsort_r?
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#define PCRE2_CODE_UNIT_WIDTH 8
#include <pcre2.h>
#include <jansson.h>

/*
 * build generated API names
 */
#ifndef CHECK_FUNCTION_NAME
#warning defining CHECK_FUNCTION_NAME macro
#define CHECK_FUNCTION_NAME check_model
#endif  // !CHECK_FUNCTION_NAME

#define concat(prefix, suffix) prefix ## suffix
#define newname(prefix, suffix) concat(prefix, suffix)

#define CHECK_map_size newname(CHECK_FUNCTION_NAME, _map_size)
#define CHECK_map_tab newname(CHECK_FUNCTION_NAME, _map_tab)
#define CHECK_fun newname(CHECK_FUNCTION_NAME, _map)
#define CHECK_init newname(CHECK_FUNCTION_NAME, _init)
#define CHECK CHECK_FUNCTION_NAME
#define CHECK_free newname(CHECK_FUNCTION_NAME, _free)

// set on init
extern char * jm_version_string;

#define array_length(arr) (sizeof(arr) / sizeof((arr)[0]))
typedef enum { false, true } bool;
typedef int (*cmp_fun_t)(const void *, const void *);
typedef int (*cmp_r_fun_t)(const void *, const void *, void *);

/*
 * JSON API extensions
 */

// keep object prop/vals as an array for sorting and fast get.
typedef struct {
    const char *prop;
    const json_t *val;
} json_propval_t;

extern bool _json_is_scalar(const json_t *);
extern int _json_cmp(const json_t *, const json_t *);
extern int _json_array_cmp(const json_t *, const json_t *);
extern int _json_object_cmp(const json_t *, const json_t *);
extern int _json_propval_cmp(const json_propval_t *, const json_propval_t *);
extern bool _json_array_unique(const json_t *);

/*
 * reporting
 */
// linked list of report entries
typedef struct ReportEntry {
  const char *message;
  char *path;
  struct ReportEntry *prev;
} ReportEntry;

typedef struct Report {
  ReportEntry *entry;
} Report;

// path segment, name == NULL means use index
typedef struct Path {
    const char *name;
    size_t index;
    struct Path *prev;  // link to previous (in stack) segment
    struct Path *next;  // used for backtacking
} Path;

extern void jm_report_add_entry(Report* rep, const char *msg, Path *path);
extern void jm_report_free_entries(Report *rep);

typedef bool (*check_fun_t)(const json_t *, Path *, Report *);

extern bool jm_check_fun_string(check_fun_t, const char *, Path *, Report *);
extern bool jm_array_is_unique(const json_t *, Path *, Report *);

/*
 * property mapping management
 */
typedef struct {
    const char *name;
    check_fun_t val_check;
} propmap_t;

extern void jm_sort_propmap(propmap_t *, int);
extern check_fun_t jm_search_propmap(const char *, const propmap_t *, int);

/*
 * set of scalar constants
 */
typedef enum {
    cst_is_null,
    cst_is_bool,
    cst_is_integer,
    cst_is_float,
    cst_is_string
    // cst_is_array + const json_t * + free
    // cst_is_object + const json_t * + free
} constant_tag;

typedef struct
{
    constant_tag tag;
    union {
        bool b;
        int64_t i;
        double f;
        const char *s;
    } val;
} constant_t;

extern bool jm_set_cst(constant_t *, const json_t *);
extern void jm_dbg_cst(const constant_t *, const char *);
extern void jm_sort_cst(constant_t *, size_t);
extern bool jm_search_cst(const constant_t *, const constant_t *, size_t);

/*
 * constant mapping management for discriminant optimization
 */
typedef struct {
    constant_t cst;
    check_fun_t check_val;
} constmap_t;

extern void jm_sort_constmap(constmap_t *, size_t);
extern check_fun_t jm_search_constmap(const constant_t *, const constmap_t *, size_t);

/*
 * Miscellaneous support functions and related data
 */
extern bool jm_is_valid_date(const char *);
extern pcre2_code *jm_is_valid_uuid_code;
extern pcre2_match_data *jm_is_valid_uuid_data;
extern bool jm_is_valid_uuid(const char *);
extern size_t jm_any_len(json_t *);
extern bool jm_is_valid_regex(const char *);
extern bool jm_is_valid_url(const char *);

typedef enum {
    op_eq,
    op_ne,
    op_le,
    op_lt,
    op_ge,
    op_gt
} constraint_op_t;

extern bool
jm_check_constraint(const json_t *, constraint_op_t, const constant_t *, Path *, Report *);

/*
 * Shared high-level entry point
 */
extern bool jm_generic_entry(char *(*)(void), check_fun_t (*)(const char *),
                             json_t *, const char *, bool *, char **);

/*
 * Generated stuff
 */
// low-level interface
extern const size_t CHECK_map_size;
extern propmap_t CHECK_map_tab[];
extern check_fun_t CHECK_fun(const char *);
// high-level interface
extern char *CHECK_init(void);
extern bool CHECK(json_t *, const char *, bool *, char **);
extern void CHECK_free(void);

#endif  // JSON_MODEL_H_INCLUDED
