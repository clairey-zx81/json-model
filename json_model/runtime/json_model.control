comment = 'JSON Model PL/pgSQL Runtime - JSON value validation'
default_version = '1.0'
