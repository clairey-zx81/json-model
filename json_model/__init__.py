#
# convenient types, classes and functions to process models
#
from .mtypes import Jsonable
from .runtime.types import Report, EntryCheckFun
from .model import JsonModel
from .script import (
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
)

__ALL__ = [
    Jsonable, Report, EntryCheckFun, JsonModel,
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
]
