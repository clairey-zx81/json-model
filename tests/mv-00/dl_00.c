#include <json-model.h>
#define JSON_MODEL_VERSION "2.0a0"

static bool json_model_2(const json_t* val, Path* path, Report* rep);
static bool json_model_1(const json_t* val, Path* path, Report* rep);
propmap_t check_model_map_tab[2];
const size_t check_model_map_size = 2;

// object $.'$rec'.'|'.1
static bool _jm_obj_0(const json_t* val, Path* path, Report* rep)
{
    if (! json_is_object(val))
    {
        if (rep) jm_report_add_entry(rep, "not an object [$.'$rec'.'|'.1]", path);
        return false;
    }
    bool res;
    const char *prop;
    json_t *pval;
    json_object_foreach((json_t *) val, prop, pval)
    {
        Path lpath_0 = (Path) { prop, 0, path, NULL };
        // handle other props
        // $.'$rec'.'|'.1.''
        res = json_model_2(pval, (path ? &lpath_0 : NULL), rep);
        if (! res)
        {
            if (rep) jm_report_add_entry(rep, "unexpected $rec [$.'$rec'.'|'.1.'']", (path ? &lpath_0 : NULL));
        }
        if (! res)
        {
            return false;
        }
    }
    return true;
}

// check $rec ($.'$rec')
static bool json_model_2(const json_t* val, Path* path, Report* rep)
{
    bool res;
    // $.'$rec'
    // $.'$rec'.'|'.0
    res = json_is_boolean(val);
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "not a bool [$.'$rec'.'|'.0]", path);
    }
    if (! res)
    {
        // $.'$rec'.'|'.1
        res = _jm_obj_0(val, path, rep);
        if (! res)
        {
            if (rep) jm_report_add_entry(rep, "not an expected object at [$.'$rec'.'|'.1]", path);
        }
    }
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "no model matched [$.'$rec'.'|']", path);
    }
    return res;
}

// check $ ($)
static bool json_model_1(const json_t* val, Path* path, Report* rep)
{
    bool res;
    // $
    res = json_model_2(val, path, rep);
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "unexpected $rec [$]", path);
    }
    return res;
}

check_fun_t check_model_map(const char *pname)
{
    return jm_search_propmap(pname, check_model_map_tab, 2);
}

static bool initialized = false;

char *CHECK_init(void)
{
    if (!initialized)
    {
        initialized = true;
        jm_version_string = JSON_MODEL_VERSION;
        check_model_map_tab[0] = (propmap_t) { "", json_model_1 };
        check_model_map_tab[1] = (propmap_t) { "rec", json_model_2 };
        jm_sort_propmap(check_model_map_tab, 2);
    }
    return NULL;
}

void CHECK_free(void)
{
    if (initialized)
    {
        initialized = false;

        // cleanup code
    }
}


/*
 * # API
 *
 * ```c
 * bool error;
 * char *message;
 *
 * bool valid = check(json_value, "model-name", &error, &message);
 *
 * if (message)
 *    free(message), message = NULL;
 * ```
 *
 * - `error` is set of there was an error during initialization or if the model does not exists.
 * - `message` is provided if a non NULL pointer is passed, and must be freed by the caller.
 */
bool
CHECK(const json_t *val, const char *name, bool *error, char **reasons)
{
    return jm_generic_entry(CHECK_init, CHECK_fun, val, name, error, reasons);
}
