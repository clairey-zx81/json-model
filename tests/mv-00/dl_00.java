//
// Generated by JSON Model Compiler version 2.0b1
// see https://github.com/clairey-zx81/json-model
//
import json_model.*;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class dl_00 extends ModelChecker
{
    static public final String VERSION = "2.0b1";

    public Map<String, Checker> dl_00_map_pmap;

    // object .'$rec'.'|'.1
    public boolean _jm_obj_0(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
        {
            if (rep != null) rep.addEntry("not an object [.'$rec'.'|'.1]", path);
            return false;
        }
        boolean res;
        Iterator<String> prop_loop = json.objectIterator(val);
        while (prop_loop.hasNext())
        {
            String prop = prop_loop.next();
            Object pval = json.objectValue(val, prop);
            Path lpath_0 = new Path(prop, path);
            // handle other props
            // .'$rec'.'|'.1.''
            res = json_model_2(pval, (path != null ? lpath_0 : null), rep);
            if (! res)
            {
                if (rep != null) rep.addEntry("unexpected $rec [.'$rec'.'|'.1.'']", (path != null ? lpath_0 : null));
            }
            if (! res)
            {
                return false;
            }
        }
        return true;
    }
    
    // check $rec (.'$rec')
    public boolean json_model_2(Object val, Path path, Report rep)
    {
        boolean res;
        // .'$rec'
        // .'$rec'.'|'.0
        res = json.isBoolean(val);
        if (! res)
        {
            if (rep != null) rep.addEntry("not a bool [.'$rec'.'|'.0]", path);
        }
        if (! res)
        {
            // .'$rec'.'|'.1
            res = _jm_obj_0(val, path, rep);
            if (! res)
            {
                if (rep != null) rep.addEntry("unexpected element [.'$rec'.'|'.1]", path);
            }
        }
        if (res)
        {
            if (rep != null) rep.clearEntries();
        }
        else
        {
            if (rep != null) rep.addEntry("no model matched [.'$rec'.'|']", path);
        }
        return res;
    }
    
    // check $ (.)
    public boolean json_model_1(Object val, Path path, Report rep)
    {
        boolean res;
        // .
        res = json_model_2(val, path, rep);
        if (! res)
        {
            if (rep != null) rep.addEntry("unexpected $rec [.]", path);
        }
        return res;
    }
    

    public void init(JSON json)
    {
        if (!initialized)
        {
            try {
            dl_00_map_pmap = new HashMap<String, Checker>();
            dl_00_map_pmap.put("", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_1(o, p, r);} });
            dl_00_map_pmap.put("rec", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_2(o, p, r);} });
                super.init(json);
            }
            catch (Exception e) {
                throw new Error("cannot initialized model checker: " + e);
            }
        }
    }

    public void free()
    {
        if (initialized)
        {
            super.free();
            dl_00_map_pmap = null;
        }
    }

    public Checker get(String name)
    {
        return dl_00_map_pmap.get(name);
    }

    static public void main(String[] args) throws Exception
    {
        ModelChecker checker = new dl_00();
        Main.main("dl_00", checker, VERSION, args);
    }
}
