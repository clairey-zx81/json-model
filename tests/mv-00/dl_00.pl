#! /usr/bin/perl
#
# Generated by JSON Model Compiler version 2
# see https://github.com/clairey-zx81/json-model
#
use strict;
use warnings;
no warnings "uninitialized";
use re::engine::RE2;
use JSON::JsonModel;
use constant JMC_VERSION => '2';


sub _jm_obj_0($$$);
sub json_model_2($$$);
sub json_model_1($$$);
my %check_model_map;

# object .'$rec'.'|'.1
sub _jm_obj_0($$$)
{
    my ($val, $path, $rep) = @_;
    if (! jm_is_object($val))
    {
        return 0;
    }
    my $res;
    scalar keys %$val;
    while (my ($prop, $pval) = each %$val)
    {
        # handle other props
        # .'$rec'.'|'.1.''
        $res = json_model_2($pval, undef, $rep);
        if (! $res)
        {
            return 0;
        }
    }
    return 1;
}

# check $rec (.'$rec')
sub json_model_2($$$)
{
    my ($val, $path, $rep) = @_;
    # .'$rec'
    # .'$rec'.'|'.0
    my $res = jm_is_boolean($val);
    if (! $res)
    {
        # .'$rec'.'|'.1
        $res = _jm_obj_0($val, $path, $rep);
    }
    return $res;
}

# check $ (.)
sub json_model_1($$$)
{
    my ($val, $path, $rep) = @_;
    # .
    return json_model_2($val, $path, $rep);
}


# initialization of global variables

our $initialized = 0;

sub check_model_init()
{
    if (!$initialized)
    {
        $initialized = 1;
        %check_model_map = (
            '' => \&json_model_2,
            'rec' => \&json_model_2,
        );
    }
}

sub check_model_free()
{
    if ($initialized)
    {
        $initialized = 0;
        %check_model_map = ();
    }
}

sub check_model($$$)
{
    my ($json, $name, $rep) = @_;
    die "unexpected model name \"$name\"" unless exists $check_model_map{$name};
    my $path = defined $rep ? [] : undef;
    return $check_model_map{$name}($json, $path, $rep);
}

#
# SCRIPT EXECUTION
#
check_model_init();
jm_main(\&check_model, \%check_model_map, JMC_VERSION);
check_model_free();
