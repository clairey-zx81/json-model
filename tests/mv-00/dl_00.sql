--
-- Generated by JSON Model Compiler version 2.0b1
-- see https://github.com/clairey-zx81/json-model
--
-- JSON_MODEL_VERSION is 2
CREATE EXTENSION IF NOT EXISTS json_model;

-- object .'$rec'.'|'.1
CREATE OR REPLACE FUNCTION _jm_obj_0(val JSONB, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  res bool;
  prop TEXT;
  pval JSONB;
BEGIN
  IF NOT (JSONB_TYPEOF(val) = 'object') THEN
    RETURN FALSE;
  END IF;
  FOR prop, pval IN SELECT * FROM JSONB_EACH(val) LOOP
    -- handle other props
    -- .'$rec'.'|'.1.''
    res := json_model_2(pval, NULL, rep);
    IF NOT res THEN
      RETURN FALSE;
    END IF;
  END LOOP;
  RETURN TRUE;
END;
$$ LANGUAGE PLpgSQL;

-- check $rec (.'$rec')
CREATE OR REPLACE FUNCTION json_model_2(val JSONB, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  res bool;
BEGIN
  -- .'$rec'
  -- .'$rec'.'|'.0
  res := JSONB_TYPEOF(val) = 'boolean';
  IF NOT res THEN
    -- .'$rec'.'|'.1
    res := _jm_obj_0(val, path, rep);
  END IF;
  RETURN res;
END;
$$ LANGUAGE PLpgSQL;

-- check $ (.)
CREATE OR REPLACE FUNCTION json_model_1(val JSONB, path TEXT[], rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  res bool;
BEGIN
  -- .
  res := json_model_2(val, path, rep);
  RETURN res;
END;
$$ LANGUAGE PLpgSQL;

CREATE OR REPLACE FUNCTION check_model_map(name TEXT)
RETURNS TEXT STRICT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  map JSONB := JSONB '{"":"json_model_1","rec":"json_model_2"}';
BEGIN
  RETURN map->>name;
END;
$$ LANGUAGE plpgsql;

--
-- constant maps initialization
--
TRUNCATE jm_constant_maps;

--
-- JSON Model checking entry point
--
-- TODO INOUT rep?
CREATE OR REPLACE FUNCTION check_model(val JSONB, name TEXT, rep jm_report_entry[])
RETURNS BOOLEAN CALLED ON NULL INPUT IMMUTABLE PARALLEL SAFE AS $$
DECLARE
  fun TEXT;
BEGIN
  fun := check_model_map(name);
  IF fun IS NULL THEN
    RAISE EXCEPTION 'model for % not found', name;
  END IF;
  RETURN jm_call(fun, val, NULL, rep);
END;
$$ LANGUAGE plpgsql;
