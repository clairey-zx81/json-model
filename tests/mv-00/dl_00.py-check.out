dl_00.values.json[0]: PASS
dl_00.values.json[1]: PASS
dl_00.values.json[2]: PASS
dl_00.values.json[3]: PASS
dl_00.values.json[4]: PASS
dl_00.values.json[5]: PASS
dl_00.values.json[6]: PASS
dl_00.values.json[7]: FAIL [("not a bool [$.'$rec'.'|'.0]", []), ("not an object [$.'$rec'.'|'.1]", []), ("not an expected object at [$.'$rec'.'|'.1]", []), ("no model matched [$.'$rec'.'|']", []), ('unexpected $rec [$]', [])]
dl_00.values.json[8]: FAIL [("not a bool [$.'$rec'.'|'.0]", []), ("not an object [$.'$rec'.'|'.1]", []), ("not an expected object at [$.'$rec'.'|'.1]", []), ("no model matched [$.'$rec'.'|']", []), ('unexpected $rec [$]', [])]
dl_00.values.json[9]: FAIL [("not a bool [$.'$rec'.'|'.0]", []), ("not an object [$.'$rec'.'|'.1]", []), ("not an expected object at [$.'$rec'.'|'.1]", []), ("no model matched [$.'$rec'.'|']", []), ('unexpected $rec [$]', [])]
dl_00.values.json[10]: FAIL [("not a bool [$.'$rec'.'|'.0]", []), ("not an object [$.'$rec'.'|'.1]", []), ("not an expected object at [$.'$rec'.'|'.1]", []), ("no model matched [$.'$rec'.'|']", []), ('unexpected $rec [$]', [])]
dl_00.values.json[11]: FAIL [("not a bool [$.'$rec'.'|'.0]", []), ("not an object [$.'$rec'.'|'.1]", []), ("not an expected object at [$.'$rec'.'|'.1]", []), ("no model matched [$.'$rec'.'|']", []), ('unexpected $rec [$]', [])]
dl_00.values.json[12]: FAIL [("not a bool [$.'$rec'.'|'.0]", []), ("not a bool [$.'$rec'.'|'.0]", []), ("not an object [$.'$rec'.'|'.1]", []), ("not an expected object at [$.'$rec'.'|'.1]", []), ("no model matched [$.'$rec'.'|']", []), ("unexpected $rec [$.'$rec'.'|'.1.'']", ['invalid']), ("not an expected object at [$.'$rec'.'|'.1]", []), ("no model matched [$.'$rec'.'|']", []), ('unexpected $rec [$]', [])]
